### PR TITLE
Use incremental resolution for minmdns discover/resolve

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
MinMDNS cannot issue AAAA queries if the "Additional data" of dnssd packets do not contain them when querying for ANY/SRV records.

#### Change overview
Use Incremental resolver from #18348 in minmdns resolver. These changes are also included:

- #18429 is contained to make 'active resolves' extensible enough to also resolve AAAA addresses
- Adds a class + unit tests to hold the target host name for AAAA. since a copy is needed, this uses heap (HeapQName.h and unit tests). We could eventually optimize for matter and use static up to 17 bytes for host name and assume always "host.local" however this is not done here.
- Adds a maximum number of parallel parsing steps configuration. Currently set to 2 to not increase RAM usage too much. 
- Removed usage of DiscoveryType from resolver (not stateful anymore).

#### Testing
This passes now:

```sh
#!/usr/bin/env bash

set -ex

export NODEID1=$RANDOM
export NODEID2=$RANDOM

echo "NODE ID USED: $NODEID1"

export THREAD_CRED=hex:0e080000000000010000000300001235060004001fffe00208c6757e4f25e551ef0708fd150126fbadd43e051083469630d05b75596a69887d5bf90c79030f4f70656e5468726561642d626231650102bb1e0410621ac8d0e53905dd1ca9b0ff94d3233d0c0402a0fff8

export CHIP_TOOL="./out/linux-x64-chip-tool/chip-tool"

$CHIP_TOOL pairing ble-thread $NODEID1 $THREAD_CRED 20202021 3840 --commissioner-name alpha
$CHIP_TOOL pairing open-commissioning-window $NODEID1 1 200 1000 3840 --commissioner-name alpha

# After commissioning
echo "Need setup payload (e.g. 34999401957)"
read setup_payload

echo "NODE ID USED: $NODEID2"

$CHIP_TOOL payload parse-setup-payload $setup_payload
$CHIP_TOOL pairing manualcode $NODEID2 $setup_payload --commissioner-name beta --ble-adapter 1234
```

CI will also validate that Discovery tests do not regress (minmdns is default on linux)
Unit tests added for the heap qname class.

### CODE REVIEW

File: src/lib/core/CHIPConfig.h [hunk 1]
```diff
@@ -1210,6 +1210,19 @@ extern const char CHIP_NON_PRODUCTION_MARKER[];
 #define CHIP_CONFIG_MINMDNS_DYNAMIC_OPERATIONAL_RESPONDER_LIST 0
 #endif // CHIP_CONFIG_MINMDNS_DYNAMIC_OPERATIONAL_RESPONDER_LIST
 
+/*
+ * @def CHIP_CONFIG_MINMDNS_MAX_PARALLEL_RESOLVES
+ *
+ * @brief Determines the maximum number of parallel SRV processing.
+ *        Affects maximum number of results received for browse requests
+ *        (where a single packet may contain multiple SRV entries)
+ *        or number of pending resolves that still require a AAAA Ip record
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPConfig.h:1219]:
```suggestion
 *        or number of pending resolves that still require a AAAA IP record
```

File: src/lib/core/CHIPConfig.h [hunk 2]
```diff
@@ -1210,6 +1210,19 @@ extern const char CHIP_NON_PRODUCTION_MARKER[];
 #define CHIP_CONFIG_MINMDNS_DYNAMIC_OPERATIONAL_RESPONDER_LIST 0
 #endif // CHIP_CONFIG_MINMDNS_DYNAMIC_OPERATIONAL_RESPONDER_LIST
 
+/*
+ * @def CHIP_CONFIG_MINMDNS_MAX_PARALLEL_RESOLVES
+ *
+ * @brief Determines the maximum number of parallel SRV processing.
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPConfig.h:1216]:
```suggestion
 * @brief Determines the maximum number of SRV records that can be processed in parallel.
```

Assuming that's accurate?

File: src/lib/dnssd/IncrementalResolve.h [hunk 3]
```diff
@@ -106,22 +106,30 @@ class IncrementalResolver
...
     RequiredInformationFlags GetMissingRequiredInformation() const;
 
-    /// Fetch the server name set by `InitializeParsing`
+    /// Fetch the target host name set by `InitializeParsing`
     ///
     /// VALIDITY: Data references internal storage of this object and is valid as long
     ///           as this object is valid and InitializeParsing is not called again.
     mdns::Minimal::SerializedQNameIterator GetTargetHostName() const { return mTargetHostName.Get(); }
 
+    /// Fetch the recrod name set by `InitializeParsing`.
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/IncrementalResolve.h:127]:
```suggestion
    /// Fetch the record name set by `InitializeParsing`.
```

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp [hunk 4]
```diff
@@ -41,338 +41,212 @@ namespace chip {
...
 constexpr size_t kMdnsMaxPacketSize = 1024;
 constexpr uint16_t kMdnsPort        = 5353;
 
 using namespace mdns::Minimal;
 
-class PacketDataReporter : public ParserDelegate
+/// Handles processing of minmdns packet data.
+///
+/// Can process multiple incremental resolves based on SRV data and allows
+/// retreival of pending (e.g. to ask for AAAA) and complete data items.
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:52]:
```suggestion
/// retrieval of pending (e.g. to ask for AAAA) and complete data items.
```

File: src/lib/support/ScopedBuffer.h [hunk 5]
```diff
@@ -142,7 +142,7 @@ class ScopedMemoryBuffer : public Impl::ScopedMemoryBufferBase<MemoryManagement>
     inline T * Get() { return static_cast<T *>(Base::Ptr()); }
     inline T & operator[](size_t index) { return Get()[index]; }
 
-    inline const T * Get() const { return static_cast<T *>(Base::Ptr()); }
+    inline const T * Get() const { return static_cast<const T *>(Base::Ptr()); }
```
Comment (bzbarsky-apple) [inline src/lib/support/ScopedBuffer.h:145]: Why this change?  Returning `T* ` from a method that returns `const T*` is fine, and casting both type and const-ness together is not great, imo....

**Review by bzbarsky-apple — APPROVED** [review]

File: src/lib/support/ScopedBuffer.h (hunk 5 above)
Comment (andy31415) [reply src/lib/support/ScopedBuffer.h:145]: Previous was not const correct: Base::Ptr() returns a const because the method is const. so static_cast to non_const was failing because it was removing constness (even though it was added right back as a return type).

**Review by tcarmelveilleux — APPROVED** [review]

File: src/lib/dnssd/ActiveResolveAttempts.cpp [hunk 6]
```diff
@@ -65,19 +65,34 @@ void ActiveResolveAttempts::Complete(const chip::Dnssd::DiscoveredNodeData & dat
...
+            return;
+        }
+    }
+}
+
 void ActiveResolveAttempts::MarkPending(const chip::PeerId & peerId)
 {
-    ScheduledAttempt attempt(peerId, /* firstSend */ true);
-    MarkPending(attempt);
+    MarkPending(ScheduledAttempt(peerId, /* firstSend */ true));
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/lib/dnssd/ActiveResolveAttempts.cpp:82]: I will do this in a followup. Also naming seems not as clear - firstSend ends up being the unicastSend flag, so we should just call it as such.
